OpenRA.Mods.TS/Traits/Render/WithVoxelWaterBody.cs

```diff
@@ -0,0 +1,81 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2015 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation. For more information,
+ * see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Graphics;
+using OpenRA.Mods.Common.Graphics;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.TS.Traits
+{
+	public class WithVoxelWaterBodyInfo : ITraitInfo, IQuantizeBodyOrientationInfo, IRenderActorPreviewVoxelsInfo, Requires<RenderVoxelsInfo>
+	{
+		public readonly string WaterSequence = "water";
+		public readonly string LandSequence = "idle";
+
+		public object Create(ActorInitializer init) { return new WithVoxelWaterBody(init.Self, this); }
+
+		public IEnumerable<VoxelAnimation> RenderPreviewVoxels(ActorPreviewInitializer init, RenderVoxelsInfo rv, string image, WRot orientation, int facings, PaletteReference p)
+		{
+			var sequence = LandSequence;
+			if (init.Contains<LocationInit>())
+			{
+				var location = init.Get<LocationInit>().Value(init.World);
+				var onWater = init.World.Map.GetTerrainInfo(location).IsWater;
+				sequence = onWater ? WaterSequence : LandSequence;
+			}
+
+			var body = init.Actor.Traits.Get<BodyOrientationInfo>();
+			var voxel = VoxelProvider.GetVoxel(image, sequence);
+			yield return new VoxelAnimation(voxel, () => WVec.Zero,
+				() => new[] { body.QuantizeOrientation(orientation, facings) },
+				() => false, () => 0);
+		}
+
+		public int QuantizedBodyFacings(ActorInfo ai, SequenceProvider sequenceProvider, string race) { return 0; }
+	}
+
+	public class WithVoxelWaterBody : IAutoSelectionSize
+	{
+		readonly Actor self;
+		readonly int2 size;
```

Comment:
I won't change this convention here. My guess this is required for `RenderSprites.AutoSelectionSize` which is a bit more complicated due to all the overlays factor into the selection size calculation.